File: lib/infrastructure/logging/core/session_manager.dart
import 'package:shared_preferences/shared_preferences.dart';
import 'package:uuid/uuid.dart';

// Responsável por dois identificadores de ciclo de vida distintos:
//
//   installation_id — UUID v4 gerado na primeira execução do app e persistido
//                     em SharedPreferences. Permanece inalterado entre sessões,
//                     upgrades e reboots. Nunca é removido por fluxos normais.
//
//   session_id      — UUID v4 gerado a cada abertura do app (em memória).
//                     Permite correlacionar eventos dentro de uma única sessão
//                     sem cruzar dados de sessões anteriores.
//
// A chave 'logger_device_id' é mantida compatível com o AppLogger legado para
// não perder o installation_id de instalações já existentes.
class SessionManager {
  SessionManager._();

  static const _installationIdKey = 'logger_device_id';
  static const _uuid = Uuid();

  static String? _installationId;
  static String? _sessionId;

  // UUID estável por instalação. Retorna '' antes de init().
  static String get installationId => _installationId ?? '';

  // UUID efêmero por sessão. Retorna '' antes de init().
  static String get sessionId => _sessionId ?? '';

  // Inicializa ambos os identificadores.
  // Deve ser chamado uma única vez no início do app (AppLogger.init()).
  static Future<void> init() async {
    _sessionId = _uuid.v4();

    final prefs = await SharedPreferences.getInstance();
    _installationId = prefs.getString(_installationIdKey);
    if (_installationId == null) {
      _installationId = _uuid.v4();
      await prefs.setString(_installationIdKey, _installationId!);
    }
  }
}

File: android/app/src/main/kotlin/com/sopro/sopro/logging/SessionManager.kt
package com.sopro.sopro.logging

import android.content.Context
import java.util.UUID

// Gerencia dois identificadores de ciclo de vida — espelho do SessionManager Flutter.
//
//   installation_id — UUID v4 gerado na primeira execução, persistido em SharedPreferences
//                     com chave 'logger_device_id'. Estável entre sessões e reboots.
//
//   session_id      — UUID v4 gerado a cada chamada a init(). Apenas em memória.
//                     Muda a cada abertura; nunca reutilizado entre execuções.
//
// Thread safety: init() usa @Synchronized. Getters leem campos @Volatile.
// Chamadas múltiplas a init() são idempotentes — apenas a primeira executa.
object SessionManager {

    private const val PREFS_NAME = "sopro_logger"
    // Chave idêntica ao Flutter para compatibilidade de installation_id entre camadas.
    private const val INSTALLATION_ID_KEY = "logger_device_id"

    @Volatile private var _installationId: String? = null
    @Volatile private var _sessionId: String? = null

    val installationId: String get() = _installationId ?: ""
    val sessionId: String get() = _sessionId ?: ""
    val isInitialized: Boolean get() = _installationId != null && _sessionId != null

    @Synchronized
    fun init(context: Context) {
        if (isInitialized) return
        _sessionId = UUID.randomUUID().toString()
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        _installationId = prefs.getString(INSTALLATION_ID_KEY, null) ?: run {
            val newId = UUID.randomUUID().toString()
            prefs.edit().putString(INSTALLATION_ID_KEY, newId).apply()
            newId
        }
        // Registra o sink Supabase após installationId estar disponível.
        // Idempotente: chamadas subsequentes de SessionManager.init() retornam antes
        // deste ponto (guard isInitialized acima), então register() roda exatamente uma vez.
        SupabaseSink.register()
    }
}
